src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,212 @@
+use crate::clean;
+use crate::html::sources;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, NestedVisitorMap, Visitor};
+use rustc_hir::{ExprKind, GenericParam, GenericParamKind, HirId, Mod, Node};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+
+use std::path::{Path, PathBuf};
+
+/// This enum allows us to store two different kinds of information:
+///
+/// In case the `span` definition comes from the same crate, we can simply get the `span` and use
+/// it as is.
+///
+/// Otherwise, we store the definition `DefId` and will generate a link to the documentation page
+/// instead of the source code directly.
+#[derive(Debug)]
+crate enum LinkFromSrc {
+    Local(clean::Span),
+    External(DefId),
+}
+
+/// This struct is used only as index in the `span_map`, not as [`Span`]! `Span`s contain
+/// some extra information (the syntax context) we don't need. **Do not convert this type back to
+/// `Span`!!!**
+#[derive(Clone, Copy, Hash, PartialEq, Eq, Debug)]
+crate struct LightSpan {
+    crate lo: u32,
+    crate hi: u32,
```

Comment:
I think it would be clearer for these fields and `file_span_lo` to be `BytePos`s rather than raw `u32`s.

Reply:
Good idea!

Reply:
Actually, maybe not: the `BytePos` type doesn't allow `+=`, so not great in some cases...

Reply:
And also, it forces to add `BytePos` wrapping a in lot of places, not making the code much easier to read...

Reply:
I still think BytePos would be better - it makes it more clear this is only used for source files, and only with valid indexes. Or is rustdoc generating invalid indexes? How can that happen?

Reply:
No, the indexes are valid, however I was complaining about the arithmetic operations which are not "in place". You always have to create a third one to store the result of the operation between two other `BytePos`. For example:

```
byte_pos += other_byte_pos; // Doesn't work!

```